expr/floatequals.go

```diff
@@ -1,6 +1,31 @@
 package expr
 
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const epsilon float64 = 0.00001
+
+// AssertFloatEquals does a fuzzy comparison of floats.
+func AssertFloatEquals(t *testing.T, a, b float64) {
+	assert.True(t, fuzzyEquals(epsilon, a, b), fmt.Sprintf("Floats did not match.  Expected: %f  Actual: %f", a, b))
+}
+
+// AssertFloatWithin checks whether a given float is within e error (decimal) of
+// another float
+func AssertFloatWithin(t *testing.T, e, expected float64, actual float64, msg string) {
+	assert.True(t, fuzzyEquals(e, expected, actual), fmt.Sprintf("%v -- Floats not within %f of each other. Expected: %f  Actual: %f", msg, e, expected, actual))
+}
+
 // courtesy of https://gist.github.com/cevaris/bc331cbe970b03816c6b
-func fuzzyEquals(epsilon, a, b float64) bool {
-	return (a-b) < epsilon && (b-a) < epsilon
+func fuzzyEquals(e, a, b float64) bool {
```

Comment:
Uses ratio tolerances now.

Reply:
Always an interesting ball of wax hehe...

So the only potential gotcha here I think is comparing to zero (or comparing two very small numbers) and how you want the tolerances to behave there...  It's debatable/kinda depends.

for example, this seems to violate my expectation:
```
fuzzyEquals(0.00001, 0.0, 1e-50) = false
```

One "fix" I've seen is to still check a bound on the absolute difference as well (like your prior version) OTOH, this case might be desirable, depending on the problem at hand (they just happen to be two small numbers, but quite different on this scale...)
```
fuzzyEquals(0.00001, 1.23456e-10,3.45678e-11) = false
```

Reply:
quick test https://gist.github.com/forkner/a8c4e249577f91c7679e1ae5f92a8d45

Reply:
Yeah, it sort of depends on what one is trying to accomplish. If the goal is purely to compare floats that are close enough within a certain floating point error, the purely subtractive comparison is more appropriate, whereas if one is trying to get a result that's "close enough" within some acceptable error from a non-precise (i.e. estimating) function, then the quotient form seems more appropriate. For now, since it's working as-is, I'm going to just leave it.